src/schema/schema_util.c

```diff
@@ -62,6 +62,48 @@ __wt_schema_get_source(WT_SESSION_IMPL *session, const char *name)
 	return (NULL);
 }
 
+/*
+ * __wt_schema_internal_session --
+ *	Create and return an internal schema session if necessary.
+ */
+int
+__wt_schema_internal_session(
+    WT_SESSION_IMPL *session, WT_SESSION_IMPL **int_sessionp)
+{
+	/*
+	 * Open an internal session if a transaction is running so that the
+	 * schema operations are not logged and buffered with any log records
+	 * in the transaction.  The new session inherits its flags from the
+	 * original.
+	 */
+	*int_sessionp = session;
+	if (F_ISSET(&session->txn, WT_TXN_RUNNING)) {
+		/* We should not have a schema txn running now. */
+		WT_ASSERT(session, !F_ISSET(session, WT_SESSION_SCHEMA_TXN));
+		WT_RET(__wt_open_internal_session(S2C(session), "schema",
+		    true, session->flags, int_sessionp));
```

Comment:
I didn't understand how `WT_TXN_RUNNING` could be set on a transaction in a new session, so I removed the code that cleared it.  (I first ran the test suite asserting that it wasn't set).  Is there some case I'm missing?

Reply:
No, I was losing a level of indirection! The session flags are set unconditionally from the original session and I was applying that in my head to the txn flags in the session too. Thanks for catching that.